sdk/go/internal/querybuilder/marshal.go

```diff
@@ -55,67 +54,63 @@ func marshalValue(ctx context.Context, v reflect.Value) (string, error) {
 
 	switch t.Kind() {
 	case reflect.Bool:
-		return fmt.Sprintf("%t", v.Bool()), nil
+		return strconv.FormatBool(v.Bool()), nil
 	case reflect.Int:
-		return fmt.Sprintf("%d", v.Int()), nil
+		return strconv.FormatInt(v.Int(), 10), nil
 	case reflect.String:
 		name := t.Name()
 		// distinguish enum const values and customScalars from string type
 		// GraphQL complains if you try to put a string literal in place of an enum: FOO vs "FOO"
 		_, found := customScalar[t.Name()]
 		if name != "string" && !found {
-			return fmt.Sprintf("%s", v.String()), nil //nolint:gosimple,staticcheck
+			return v.String(), nil
 		}
-		return fmt.Sprintf("%q", v.String()), nil //nolint:gosimple,staticcheck
+		return strconv.Quote(v.String()), nil
 	case reflect.Pointer:
 		if v.IsNil() {
 			return "null", nil
 		}
 		return marshalValue(ctx, v.Elem())
 	case reflect.Slice:
+		var b strings.Builder
+		b.WriteRune('[')
 		n := v.Len()
-		elems := make([]string, n)
-		eg, gctx := errgroup.WithContext(ctx)
 		for i := 0; i < n; i++ {
-			i := i
-			eg.Go(func() error {
```

Comment:
Sorry, totally missed this comment.

It's because of "pass by object".

When passing a `*Container` type, there's a custom marshaller doing the `*Container` --> `ContainerID` conversion which issues an API call: https://github.com/dagger/dagger/blob/main/sdk/go/api.gen.go#L79

The problem is worsened since this is a recursive: passing an array with 10 elements, each containing a struct with 10 fields will result in 100 sequential API calls. On the upside, I don't *think* the API currently has a scenario where that is possible.

There are two layers of parallelism:
- `querybuilder.go:marshalArguments()`: For a given GraphQL call, process all arguments in parallel (e.g. if chaining 10 things -- take all passed arguments and resolve them at once)
- *this code*: for each argument, resolve each slice and struct elements in parallel

Combined they make sure that all *Object->ObjectID resolution happens in parallel

Reply:
@aluzzardi Ah - I didn't appreciate that marshaling involves sending an API request, that makes sense. Sounds important; will revert and reopen.